### PR TITLE
Add conversation tab

### PR DESCRIPTION
**Pull request by JoshLove-msft**
Addresses https://github.com/Azure/azure-sdk-tools/issues/730
Shows comment threads grouped by the latest revision that a comment was added. When clicking the revision title link, the section will collapse. Clicking the element Id will jump to the comment in the corresponding revision in the Api tab in a new browser tab.

image

### CODE REVIEW

**Comment by pakrym** [comment]
Nice!

Can we reuse the comment thread template from the API page and format member name as a code line?

**Comment by JoshLove-msft** [comment]
> Nice!
> 
> Can we reuse the comment thread template from the API page and format member name as a code line?

I can update the formatting, but in terms of the the thread template I was thinking this would be a read-only view into the comments. Were you thinking that users should actually be able to act on the comments from here?

**Comment by pakrym** [comment]
> Were you thinking that users should actually be able to act on the comments from here?

If it's not hard to do would be cool to have the entire functionality.

**Comment by JoshLove-msft** [comment]
> > Were you thinking that users should actually be able to act on the comments from here?
> 
> If it's not hard to do would be cool to have the entire functionality.

Eh I'm not so sure about this. There could be other context from the review that would only be visible from the API tab. Also, it will end up taking up way more real estate to use the thread template. I think of the Summary tab as basically a report showing read-only info for the review.

**Comment by pakrym** [comment]
>There could be other context from the review that would only be visible from the API tab. 

Like what?

**Comment by pakrym** [comment]
I want to eventually have something like this:

image

**Comment by JoshLove-msft** [comment]
> > There could be other context from the review that would only be visible from the API tab.
> 
> Like what?

If the comment references other elements than the one that is being commented on. I can try it out though and see how it looks.

File: src/dotnet/APIView/APIViewWeb/Pages/Assemblies/Summary.cshtml [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+        }
+        else
+        {
+            <div class="p-3">
+                <div class="card border-primary">
+                    @for (int i = 0; i < 2; i++)
+                    {
+                        string threadDisplay = "";
+                        string divId = "";
+                        IOrderedEnumerable<KeyValuePair<ReviewRevisionModel, List<APIViewWeb.Models.CommentThreadModel>>> threads;
```
Comment (pakrym) [inline src/dotnet/APIView/APIViewWeb/Pages/Assemblies/Summary.cshtml:21]: Can we throw this logic into the model?

**Comment by JoshLove-msft** [comment]
> > > There could be other context from the review that would only be visible from the API tab.
> > 
> > 
> > Like what?
> 
> If the comment references other elements than the one that is being commented on. I can try it out though and see how it looks.

@tg-msft, as the initial requester of this feature, do you have thoughts on this?

**Comment by tg-msft** [comment]
Here's how I see the world:

> (right now) < (just links) < (links + comment text) < (links + editable comments)

It would be way less frustrating to have comment summaries rather than click the link, realize it was the wrong one, click a different link, realize it's still the wrong one, click a third link, realize it's still the wrong one, and finally end up weeping into my cat.

I'm beyond happy with any forward progress though and we can do this in stages.

**Comment by JoshLove-msft** [comment]
> Here's how I see the world:
> 
> > (right now) < (just links) < (links + comment text) < (links + editable comments)
> 
> It would be way less frustrating to have comment summaries rather than click the link, realize it was the wrong one, click a different link, realize it's still the wrong one, click a third link, realize it's still the wrong one, and finally end up weeping into my cat.
> 
> I'm beyond happy with any forward progress though and we can do this in stages.

Currently, the comment is shown and links out to the comment in the revision that it is in. I am going to investigate adding the full editable comments since we already have all the code for this it is just a matter of loading in all the data that we need.

**Comment by tg-msft** [comment]
To be clear, are we talking about "editing my existing comment" or "replying to an existing comment"?  I'd like to be able to quickly reply or mark things resolved if that were possible.  I care a lot less about editing my own comments from that view.

**Comment by JoshLove-msft** [comment]
> To be clear, are we talking about "editing my existing comment" or "replying to an existing comment"? I'd like to be able to quickly reply or mark things resolved if that were possible. I care a lot less about editing my own comments from that view.

Well we would get the whole markup so it also includes the ability to edit (and the client side scripts should already be loaded). We can of course remove that when displayed in this other tab, but I don't think it would be extra work to include it.

File: src/dotnet/APIView/APIViewWeb/Pages/Assemblies/Conversation.cshtml [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+                                    <button class="btn btn-link ml-n3" data-toggle="collapse" data-target="#@divId" aria-expanded="true" aria-controls=@divId>
+                                        @revision.Key.DisplayName
+                                    </button>
+                                </h5>
+                            </div>
+                            <div id=@divId class="collapse show" data-revision-id=@revision.Key.RevisionId aria-labelledby="header-@revision.Key.RevisionId">
+                                @foreach (var thread in revision.Value)
+                                {
+                                    var elementId = @thread.Comments.First().ElementId;
+                                    var uri = new Uri($"{Model.Endpoint}/Assemblies/Review/{Model.Review.ReviewId}/{revision.Key.RevisionId}#{Uri.EscapeUriString(elementId)}");
```
Comment (pakrym) [inline src/dotnet/APIView/APIViewWeb/Pages/Assemblies/Conversation.cshtml:32]: Use this https://docs.microsoft.com/en-us/aspnet/core/mvc/views/tag-helpers/built-in/anchor-tag-helper?view=aspnetcore-3.1

File: src/dotnet/APIView/APIViewWeb/Pages/Assemblies/Conversation.cshtml [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+                                <h5 class="mb-0">
+                                    <button class="btn btn-link ml-n3" data-toggle="collapse" data-target="#@divId" aria-expanded="true" aria-controls=@divId>
+                                        @revision.Key.DisplayName
+                                    </button>
+                                </h5>
+                            </div>
+                            <div id=@divId class="collapse show" data-revision-id=@revision.Key.RevisionId aria-labelledby="header-@revision.Key.RevisionId">
+                                @foreach (var thread in revision.Value)
+                                {
+                                    var elementId = @thread.Comments.First().ElementId;
```
Comment (pakrym) [inline src/dotnet/APIView/APIViewWeb/Pages/Assemblies/Conversation.cshtml:31]: Isn't https://github.com/Azure/azure-sdk-tools/blob/master/src/dotnet/APIView/APIViewWeb/Models/CommentThreadModel.cs#L20 an ElementId?

**Comment by JoshLove-msft** [comment]
> Isn't it a beauty?

Ideally the comment boxes would all be the same width.

**Comment by JoshLove-msft** [comment]
> > Isn't it a beauty?
> 
> Ideally the comment boxes would all be the same width.

Fixed

**Pull request merged**